clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/cache/CatsOnDemandCacheUpdater.groovy

```diff
@@ -57,13 +61,27 @@ class CatsOnDemandCacheUpdater implements OnDemandCacheUpdater {
   OnDemandCacheUpdater.OnDemandCacheStatus handle(OnDemandAgent.OnDemandType type, Collection<OnDemandAgent> onDemandAgents, Map<String, ? extends Object> data) {
     boolean hasOnDemandResults = false
     for (OnDemandAgent agent : onDemandAgents) {
+      boolean atomicScheduler = false;
+      AgentScheduler scheduler = null
+      if (agent instanceof AgentSchedulerAware) {
+        scheduler = agent.agentScheduler
+        atomicScheduler = scheduler.atomic
+      }
       try {
+        AgentLock lock = null;
+        if (atomicScheduler && !(lock = scheduler.tryLock((Agent) agent))) {
+          hasOnDemandResults = true // force Orca to retry
+          continue;
+        }
         final long startTime = System.nanoTime()
         def providerCache = catsModule.getProviderRegistry().getProviderCache(agent.providerName)
         OnDemandAgent.OnDemandResult result = agent.handle(providerCache, data)
         if (result) {
+          if (atomicScheduler && !(scheduler.tryRelease(lock))) {
+            hasOnDemandResults = true // force Orca to retry
+          }
           if (result.cacheResult) {
-            hasOnDemandResults = !(result.cacheResult.cacheResults ?: [:]).values().flatten().isEmpty()
+            hasOnDemandResults = !(result.cacheResult.cacheResults ?: [:]).values().flatten().isEmpty() && !atomicScheduler
             agent.metricsSupport.cacheWrite {
```

Comment:
It's possible that 
```
assert(lock.isValid())
assert(!lock.isValid())
```
passes, but I guess that's even less likely than either of the above scenarios. I'll go with the `isValid()`.